### PR TITLE
Added support for virtual hosts buckets

### PR DESCRIPTION
**Pull request by lackas**
which allows transparent access to buckets outside the US-Standard region
by looking up the region via DNS.

Rusty, this looks a bit like a hack. However, I have put some thought into this, e.g. adjusting the path method to not return the bucket name, and rather pass the bucketname into the constructor also. That said, this approach has the least impact on anything and should work reliably.

It does work in my test environment, where I can now switch between buckets of various regions without any changes to the host.

### CODE REVIEW

**Comment by rustyconover** [comment]
@lackas Why not prefer this method when the bucket name can be reliably used as a DNS based virtual name than making the user worry about it?  If the bucket name can be used as a virtual host name, we can eliminate the configuration and just go with convention.  If a user wants to specify the API endpoint we should still provide them a way to do that but that wouldn't be the default use case.  What do you think?

**Comment by lackas** [comment]
Thanks for the feedback. The virtual host setup would be our default and also Amazon sees it as the preferred method these days. However, your module most likely does not have many non-US users (or they have to deal with the regions manually). So changing the default (or even getting rid of the old path-style) would break at least two things:
- buckets with non-DNS compliant names, which cannot be used in the host name
- secure connection to buckets with a dot in their name (the AWS certificate only covers *.s3.amazonaws.com, so something like bucket.company.s3.amazonaws.com would raise an error in LWP; which one theoretically could catch and validate separately)
Long story short, I believe it's best to support both methods. I would like to see virtual host become the default, however, maybe at a later point in time when people got more used to it.
